Missing sources variants in emitted .module, causing dependency locking failure

### Background

POM files *can* explicitly reference sources or javadoc variants, but from what I've seen, they never do.

Gradle's approach to dependency resolution, when the package source turns out to have a POM file and no Gradle `.module` file, is (to my understanding) to assume that the POM won't bother to mention `-sources` or `-javadoc` variants even if they do exist. Instead, Gradle does its own heuristic: if it's considering a POM the source-of-truth for a package, then it just builds URLs for `-sources` and `-javadoc` `.jar` artifact files; if it finds them to be there, then it synthesizes variants for these. If they're not there, it doesn't.

On the other hand, if you generate a Gradle `.module` file, Gradle assumes you know what you're doing, and doesn't do this heuristic. It expects you to explicitly list any sources variants in the module file.

[`kotlin-stdlib` has an example of a `.module` file](https://repo1.maven.org/maven2/org/jetbrains/kotlin/kotlin-stdlib/2.2.0/kotlin-stdlib-2.2.0.module) where the sources variants are explicitly referenced:

```json
    {
      "name": "jvmSourcesElements",
      "attributes": {
        "org.gradle.category": "documentation",
        "org.gradle.dependency.bundling": "external",
        "org.gradle.docstype": "sources",
        "org.gradle.jvm.environment": "standard-jvm",
        "org.gradle.libraryelements": "jar",
        "org.gradle.usage": "java-runtime",
        "org.jetbrains.kotlin.platform.type": "jvm"
      },
      "files": [
        {
          "name": "kotlin-stdlib-2.2.0-sources.jar",
          "url": "kotlin-stdlib-2.2.0-sources.jar",
          "size": 698989,
          "sha512": "b41875f498f33f2687bd07dfc1cbbf5504454eba9a7a78cac77948009ac67272a1476689524ad8f31aa9d8c661be66d6eefb42a8dfb0cc0309ef3bc8b64b54ba",
          "sha256": "967ad9599254e3a60d96d6c789547cc35c22d770d9c8fb1e3f15fac3b4c3b65d",
          "sha1": "a2596f4aca441c6ba0ad915501c2bfa2e3c51c0b",
          "md5": "b736e8046405648ebf5639f17cbd3d8b"
        }
      ]
    }
```

`jackson-core`, on the other hand, is an example of a Maven project that uses this project as a plugin to support adding additional metadata for Gradle dependents. `jackson-core` ends up *without* these sources variants in its `.module` file, despite publishing `-sources` and `-javadoc` jars to MCR. [Here's their `.module` file](https://repo1.maven.org/maven2/com/fasterxml/jackson/core/jackson-core/2.19.1/jackson-core-2.19.1.module), for reference.

### Problem

Without these sources-variant stanzas in the `.module` file, some parts of Gradle seem to still assume that these variants exist as long as the artifacts do—and so Gradle adds references to these variants to its package dependency graph. But whenever Gradle actually runs through resolution of its package dependency graph, that resolution process *partially fails*.

```
$ ./gradlew dependencies --write-locks --info
...
dependencySources
Variant Selection Exception: org.gradle.internal.component.resolution.failure.exception.VariantSelectionByAttributesException caused by Resolution Failure: org.gradle.internal.component.resolution.failure.type.NoCompatibleVariantsFailure
...
+--- com.fasterxml.jackson.core:jackson-core:2.18.3 FAILED
...
BUILD SUCCESSFUL in 375ms
```

While this command succeeds in creating a `gradle.lockfile`, the resulting `gradle.lockfile` contains no entries for the referenced `dependencySources` configuration. This essentially means that a *partial* or *corrupt* `gradle.lockfile` has been created.

* Gradle writing out a lockfile in this circumstance might be considered a gradle bug, depending on your perspective. (The resulting lockfile still works... but only for some commands, and only in non-strict lock mode.)
* Interestingly, this happens despite the variant's artifact being successfully downloaded into the Gradle cache. Gradle knows it wants the variant/artifact; it even manages to *retrieve* the variant's artifact; but then it essentially "loses track of it".
* Oddly enough, you can explicitly depend on the source variant yourself (e.g. `implementation("foo:bar:0.1.0:sources")` — and this explicit resolution works (probably because, when you specify the variant name, you skip the `VariantSelectionByAttributes` search.) However, this does nothing to help the implicit resolution succeed. Even if you explicitly depend on all the same variants, the *implicit* dependencies still exist in the dependency graph, and fail to resolve, so the resulting lockfile is still partial/corrupt from Gradle's perspective.

If the dependent project's developer has activated strict dependency-locking in their project's `build.gradle.kts`, then many gradle commands will now fail:

```
$ ./gradlew dependencies
dependencySources

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':dependencies'.
> Could not resolve all dependencies for configuration ':dependencySources'.
   > Locking strict mode: Configuration ':dependencySources' is locked but does not have lock state.
```

No amount of public-API Gradle commands, nor even black-magic Gradle-internal-API voodoo in custom Gradle tasks, will force these metadata-deficient sources variants to derive and lock in the lockfile.

### Proposed Solution

Match Gradle's dependency-fetch-time POM variant-deduction heuristic, at package-publication `.module`-file-generation time.

That is: in any situation where Gradle would inject/synthesize sources/javadoc variants in memory for a package, this generator should synthesize sources/javadoc variants and write them into the `.module` it's emitting.

--- a/src/main/java/org/gradlex/maven/gmm/GradleModuleMetadataWriter.java
+++ b/src/main/java/org/gradlex/maven/gmm/GradleModuleMetadataWriter.java
@@ -30,6 +30,7 @@ import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.TreeMap;
 
 /**
@@ -40,16 +41,24 @@ public class GradleModuleMetadataWriter {
     private static final String FORMAT_VERSION = "1.1";
 
     private enum Variant {
-        API_ELEMENTS("apiElements", "java-api", Collections.singletonList("compile")),
-        RUNTIME_ELEMENTS("runtimeElements", "java-runtime", Arrays.asList("compile", "runtime"));
+        API_ELEMENTS("apiElements", "java-api", "library", "jar", null, Collections.singletonList("compile")),
+        RUNTIME_ELEMENTS("runtimeElements", "java-runtime", "library", "jar", null, Arrays.asList("compile", "runtime")),
+        JAVADOC_ELEMENTS("javadocElements", "java-runtime", "documentation", null, "javadoc", Collections.emptyList()),
+        SOURCES_ELEMENTS("sourcesElements", "java-runtime", "documentation", null, "sources", Collections.emptyList());
 
         private final String name;
         private final String usage;
+        private final String category;
+        private final String libraryelements; //nullable
+        private final String docstype; // nullable
         private final List<String> scopes;
 
-        Variant(String name, String usage, List<String> scopes) {
+        Variant(String name, String usage, String category, String libraryelements, String docstype, List<String> scopes) {
             this.name = name;
             this.usage = usage;
+            this.category = category;
+            this.libraryelements = libraryelements;
+            this.docstype = docstype;
             this.scopes = scopes;
         }
     }
@@ -78,9 +87,14 @@ public class GradleModuleMetadataWriter {
     private static Map<String, String> variantAttributes(Variant variant) {
         Map<String, String> attributes = new TreeMap<>();
 
-        attributes.put("org.gradle.category", "library");
+        attributes.put("org.gradle.category", variant.category);
         attributes.put("org.gradle.dependency.bundling", "external");
-        attributes.put("org.gradle.libraryelements", "jar");
+        if (variant.libraryelements != null) {
+            attributes.put("org.gradle.libraryelements", variant.libraryelements);
+        }
+        if (variant.docstype != null) {
+            attributes.put("org.gradle.docstype", variant.docstype);
+        }
 
         attributes.put("org.gradle.usage", variant.usage);
 
@@ -124,10 +138,20 @@ public class GradleModuleMetadataWriter {
                                       List<Dependency> removedDependencies,
                                       List<Dependency> compileOnlyApiDependencies,
                                       JsonWriter jsonWriter) throws IOException {
+
+        Optional<Artifact> javadocJar = project.getAttachedArtifacts().stream().filter(jar -> "javadoc".equals(jar.getClassifier())).findFirst();
+        Optional<Artifact> sourcesJar = project.getAttachedArtifacts().stream().filter(jar -> "sources".equals(jar.getClassifier())).findFirst();
+
         jsonWriter.name("variants");
         jsonWriter.beginArray();
-        writeVariant(project, Variant.API_ELEMENTS, platformDependencies, capabilities, removedDependencies, compileOnlyApiDependencies, jsonWriter);
-        writeVariant(project, Variant.RUNTIME_ELEMENTS, platformDependencies, capabilities, removedDependencies, null, jsonWriter);
+        writeVariant(project, Variant.API_ELEMENTS, platformDependencies, capabilities, removedDependencies, compileOnlyApiDependencies, project.getArtifact(), jsonWriter);
+        writeVariant(project, Variant.RUNTIME_ELEMENTS, platformDependencies, capabilities, removedDependencies, null, project.getArtifact(), jsonWriter);
+        if (javadocJar.isPresent()) {
+            writeVariant(project, Variant.JAVADOC_ELEMENTS, null, null, null, null, javadocJar.get(), jsonWriter);
+        }
+        if (sourcesJar.isPresent()) {
+            writeVariant(project, Variant.SOURCES_ELEMENTS, null, null, null, null, sourcesJar.get(), jsonWriter);
+        }
         jsonWriter.endArray();
     }
 
@@ -141,13 +165,13 @@ public class GradleModuleMetadataWriter {
                                      List<Capability> capabilities,
                                      List<Dependency> removedDependencies,
                                      List<Dependency> addedDependencies,
-                                     JsonWriter jsonWriter) throws IOException {
+                                     Artifact artifact, JsonWriter jsonWriter) throws IOException {
         jsonWriter.beginObject();
         jsonWriter.name("name");
         jsonWriter.value(variant.name);
         writeAttributes(variantAttributes(variant), jsonWriter);
         writeDependencies(variant, project.getDependencies(), platformDependencies, removedDependencies, addedDependencies, jsonWriter);
-        writeArtifacts(project, jsonWriter);
+        writeArtifacts(artifact, jsonWriter);
         writeCapabilities(project, capabilities, jsonWriter);
 
         jsonWriter.endObject();
@@ -168,10 +192,10 @@ public class GradleModuleMetadataWriter {
         jsonWriter.endObject();
     }
 
-    private static void writeArtifacts(MavenProject project, JsonWriter jsonWriter) throws IOException {
+    private static void writeArtifacts(Artifact artifact, JsonWriter jsonWriter) throws IOException {
         jsonWriter.name("files");
         jsonWriter.beginArray();
-        writeArtifact(project.getArtifact(), jsonWriter);
+        writeArtifact(artifact, jsonWriter);
         jsonWriter.endArray();
     }
 

--- a/src/test/java/org/gradlex/maven/gmm/test/GMMMavenPluginTest.java
+++ b/src/test/java/org/gradlex/maven/gmm/test/GMMMavenPluginTest.java
@@ -31,6 +31,7 @@ import java.io.FileReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.Arrays;
 import java.util.List;
 
@@ -246,6 +247,21 @@ class GMMMavenPluginTest {
         assertExpectedGMM("variant-dependencies");
     }
 
+    @Test
+    void testWithJavadoc() {
+        assertExpectedGMM("with-javadoc");
+    }
+
+    @Test
+    void testWithSources() {
+        assertExpectedGMM("with-sources");
+    }
+
+    @Test
+    void testWithJavadocAndSources() {
+        assertExpectedGMM("with-javadoc-and-sources");
+    }
+
     List<String> resolve() {
         BuildResult buildResult = GradleRunner.create()
                 .forwardOutput()
@@ -280,6 +296,7 @@ class GMMMavenPluginTest {
         File testPom = new File("src/test/resources/" + name + "/pom.xml");
         File testPomParent = new File("src/test/resources/" + name + "/parent/pom.xml");
         File gmmExpected = new File("src/test/resources/" + name + "/expected-module.json");
+        File javaClass = new File("src/test/resources/" + name + "/Dummy.java");
         assertThat(gmmExpected).exists();
 
         try {
@@ -289,6 +306,11 @@ class GMMMavenPluginTest {
                 Files.createDirectories(mavenProducerParent.getParentFile().toPath());
                 Files.copy(testPomParent.toPath(), mavenProducerParent.toPath());
             }
+            if (javaClass.exists()) {
+                Path target = mavenProducerBuild.getParentFile().toPath().resolve("src/main/java/org/example/Dummy.java");
+                Files.createDirectories(target.getParent());
+                Files.copy(javaClass.toPath(), target);
+            }
 
             packageProducer();
 
@@ -297,7 +319,7 @@ class GMMMavenPluginTest {
 
             JsonElement expected = JsonParser.parseReader(new FileReader(gmmExpected));
             JsonElement actual = JsonParser.parseReader(new FileReader(gmmActual));
-            assertThat(expected).isEqualTo(actual);
+            assertThat(actual).isEqualTo(expected);
         } catch (IOException e) {
             throw new RuntimeException(e);
         }

--- a/src/test/resources/with-javadoc-and-sources/Dummy.java
+++ b/src/test/resources/with-javadoc-and-sources/Dummy.java
@@ -1,0 +1,6 @@
+package org.example;
+
+/**
+ * Doc
+ */
+public class Dummy {}

--- a/src/test/resources/with-javadoc-and-sources/expected-module.json
+++ b/src/test/resources/with-javadoc-and-sources/expected-module.json
@@ -1,0 +1,119 @@
+{
+  "formatVersion": "1.1",
+  "component": {
+    "group": "org.testing",
+    "module": "example",
+    "version": "1.0",
+    "attributes": {
+      "org.gradle.status": "release"
+    }
+  },
+  "variants": [
+    {
+      "name": "apiElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-api"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2182,
+          "sha512": "e3904e23db50e6d095b5f3f504c1f5da168341b6221584c8b060e8a29d6c632ad5936bed252d33aa0856f9de54300619a153fa5050432dcc20f1699c8d2087a5",
+          "sha256": "6576b82592def810f66407778861e13a14dfc2bca37be7510e49c2e9bf983d6e",
+          "sha1": "49ccac69a7c955fd2fc88ea49dfcc36d7f1ef044",
+          "md5": "40241c3e9ff52d412e298b11b107fbfc"
+        }
+      ]
+    },
+    {
+      "name": "runtimeElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-runtime"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2182,
+          "sha512": "e3904e23db50e6d095b5f3f504c1f5da168341b6221584c8b060e8a29d6c632ad5936bed252d33aa0856f9de54300619a153fa5050432dcc20f1699c8d2087a5",
+          "sha256": "6576b82592def810f66407778861e13a14dfc2bca37be7510e49c2e9bf983d6e",
+          "sha1": "49ccac69a7c955fd2fc88ea49dfcc36d7f1ef044",
+          "md5": "40241c3e9ff52d412e298b11b107fbfc"
+        }
+      ]
+    },
+    {
+      "name": "javadocElements",
+      "attributes": {
+        "org.gradle.category": "documentation",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.docstype": "javadoc",
+        "org.gradle.usage": "java-runtime"
+      },
+      "files": [
+        {
+          "name": "example-1.0-javadoc.jar",
+          "url": "example-1.0-javadoc.jar",
+          "size": 82294,
+          "sha512": "1aeb842140bbd05bbfba4883826fb5f14a7d5edce2af77304c4d321346864abe617137576e85201813e128e6b9a5e953730b5b9456b7b0a6ad370dbeee9b09e3",
+          "sha256": "a6230826fb438bb6ee392295e4e1901fbe92b6d854dd7c72f421f694aa1bc1e8",
+          "sha1": "071a6ac4c28b8b573cb647743c37953a4c9c9dba",
+          "md5": "7daa17d2c9da4a70c984ee39ef05f6f6"
+        }
+      ]
+    },
+    {
+      "name": "sourcesElements",
+      "attributes": {
+        "org.gradle.category": "documentation",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.docstype": "sources",
+        "org.gradle.usage": "java-runtime"
+      },
+      "files": [
+        {
+          "name": "example-1.0-sources.jar",
+          "url": "example-1.0-sources.jar",
+          "size": 2018,
+          "sha512": "49453629e5b110824b33adfff1527ccbdb79b1b0c958f4f15e919a15584e97e8923370fa32432e1a723acb34c5a9e6f8b831be54cc6d8a97d20a6a8258be9c17",
+          "sha256": "1f20f7ba9c82e12187f61bbbfa232ddc99c8595c1c0ff055a6e5a3b652af1be8",
+          "sha1": "b5e2375cb759a08bc638761220a756f6fad16ea5",
+          "md5": "abfbaffd482cd013e146031fae07660a"
+        }
+      ]
+    }
+  ]
+}

--- a/src/test/resources/with-javadoc-and-sources/pom.xml
+++ b/src/test/resources/with-javadoc-and-sources/pom.xml
@@ -1,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <!-- do_not_remove: published-with-gradle-metadata -->
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.testing</groupId>
+  <artifactId>example</artifactId>
+  <version>1.0</version>
+  <packaging>jar</packaging>
+  <name>Test GMM Mojo</name>
+  <properties>
+    <maven.compiler.release>8</maven.compiler.release>
+    <project.build.outputTimestamp>2023-01-01T00:00:00Z</project.build.outputTimestamp>
+  </properties>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-javadoc-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>attach-javadocs</id>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-source-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>attach-sources</id>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.gradlex</groupId>
+        <artifactId>gradle-module-metadata-maven-plugin</artifactId>
+        <executions>
+          <execution>
+            <goals>
+              <goal>gmm</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <platformDependencies>
+            <dependency>
+              <groupId>org.junit</groupId>
+              <artifactId>junit-bom</artifactId>
+              <version>5.6.0</version>
+            </dependency>
+          </platformDependencies>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/src/test/resources/with-javadoc/Dummy.java
+++ b/src/test/resources/with-javadoc/Dummy.java
@@ -1,0 +1,6 @@
+package org.example;
+
+/**
+ * Doc
+ */
+public class Dummy {}

--- a/src/test/resources/with-javadoc/expected-module.json
+++ b/src/test/resources/with-javadoc/expected-module.json
@@ -1,0 +1,99 @@
+{
+  "formatVersion": "1.1",
+  "component": {
+    "group": "org.testing",
+    "module": "example",
+    "version": "1.0",
+    "attributes": {
+      "org.gradle.status": "release"
+    }
+  },
+  "variants": [
+    {
+      "name": "apiElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-api"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2167,
+          "sha512": "46f5f37ff29799cf65b1500ddbeb3c7bd280611b8ebb14942afc8c4884ef34e044a71d6a4ee225131ef1f5a815765bada75578d4bfcf3357503042768e91774f",
+          "sha256": "f2d928668696b8e8883b899f440eb562f22b6ae38469bc7e309beefba658c504",
+          "sha1": "a8ae492a465d3cffe0bcc16bc459678520f99bbf",
+          "md5": "780f949b1baefbaaac2a4113a8f6d6c5"
+        }
+      ]
+    },
+    {
+      "name": "runtimeElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-runtime"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2167,
+          "sha512": "46f5f37ff29799cf65b1500ddbeb3c7bd280611b8ebb14942afc8c4884ef34e044a71d6a4ee225131ef1f5a815765bada75578d4bfcf3357503042768e91774f",
+          "sha256": "f2d928668696b8e8883b899f440eb562f22b6ae38469bc7e309beefba658c504",
+          "sha1": "a8ae492a465d3cffe0bcc16bc459678520f99bbf",
+          "md5": "780f949b1baefbaaac2a4113a8f6d6c5"
+        }
+      ]
+    },
+    {
+      "name": "javadocElements",
+      "attributes": {
+        "org.gradle.category": "documentation",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.docstype": "javadoc",
+        "org.gradle.usage": "java-runtime"
+      },
+      "files": [
+        {
+          "name": "example-1.0-javadoc.jar",
+          "url": "example-1.0-javadoc.jar",
+          "size": 82294,
+          "sha512": "1aeb842140bbd05bbfba4883826fb5f14a7d5edce2af77304c4d321346864abe617137576e85201813e128e6b9a5e953730b5b9456b7b0a6ad370dbeee9b09e3",
+          "sha256": "a6230826fb438bb6ee392295e4e1901fbe92b6d854dd7c72f421f694aa1bc1e8",
+          "sha1": "071a6ac4c28b8b573cb647743c37953a4c9c9dba",
+          "md5": "7daa17d2c9da4a70c984ee39ef05f6f6"
+        }
+      ]
+    }
+  ]
+}

--- a/src/test/resources/with-javadoc/pom.xml
+++ b/src/test/resources/with-javadoc/pom.xml
@@ -1,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <!-- do_not_remove: published-with-gradle-metadata -->
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.testing</groupId>
+  <artifactId>example</artifactId>
+  <version>1.0</version>
+  <packaging>jar</packaging>
+  <name>Test GMM Mojo</name>
+  <properties>
+    <maven.compiler.release>8</maven.compiler.release>
+    <project.build.outputTimestamp>2023-01-01T00:00:00Z</project.build.outputTimestamp>
+  </properties>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-javadoc-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>attach-javadocs</id>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.gradlex</groupId>
+        <artifactId>gradle-module-metadata-maven-plugin</artifactId>
+        <executions>
+          <execution>
+            <goals>
+              <goal>gmm</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <platformDependencies>
+            <dependency>
+              <groupId>org.junit</groupId>
+              <artifactId>junit-bom</artifactId>
+              <version>5.6.0</version>
+            </dependency>
+          </platformDependencies>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

--- a/src/test/resources/with-sources/Dummy.java
+++ b/src/test/resources/with-sources/Dummy.java
@@ -1,0 +1,6 @@
+package org.example;
+
+/**
+ * Doc
+ */
+public class Dummy {}

--- a/src/test/resources/with-sources/expected-module.json
+++ b/src/test/resources/with-sources/expected-module.json
@@ -1,0 +1,99 @@
+{
+  "formatVersion": "1.1",
+  "component": {
+    "group": "org.testing",
+    "module": "example",
+    "version": "1.0",
+    "attributes": {
+      "org.gradle.status": "release"
+    }
+  },
+  "variants": [
+    {
+      "name": "apiElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-api"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2166,
+          "sha512": "8bf3027fffce66a8482299be1e663727e71639411a8cb0140187d353fd833e2df76af33dcf2b6660f6a4565e40460d361630d83f98bd01bad1885c0f160591a5",
+          "sha256": "1fc5510dfc8fdf1e095bd521250b387d5a8de991d0825e7a35dbb116d6ba1dcb",
+          "sha1": "2036458ad3ace5f93c9f82b0b80abb3d2299ebd1",
+          "md5": "53c510f182842b10afa3aab503dd6bd1"
+        }
+      ]
+    },
+    {
+      "name": "runtimeElements",
+      "attributes": {
+        "org.gradle.category": "library",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.libraryelements": "jar",
+        "org.gradle.usage": "java-runtime"
+      },
+      "dependencies": [
+        {
+          "group": "org.junit",
+          "module": "junit-bom",
+          "version": {
+            "requires": "5.6.0"
+          },
+          "attributes": {
+            "org.gradle.category": "platform"
+          },
+          "endorseStrictVersions": true
+        }
+      ],
+      "files": [
+        {
+          "name": "example-1.0.jar",
+          "url": "example-1.0.jar",
+          "size": 2166,
+          "sha512": "8bf3027fffce66a8482299be1e663727e71639411a8cb0140187d353fd833e2df76af33dcf2b6660f6a4565e40460d361630d83f98bd01bad1885c0f160591a5",
+          "sha256": "1fc5510dfc8fdf1e095bd521250b387d5a8de991d0825e7a35dbb116d6ba1dcb",
+          "sha1": "2036458ad3ace5f93c9f82b0b80abb3d2299ebd1",
+          "md5": "53c510f182842b10afa3aab503dd6bd1"
+        }
+      ]
+    },
+    {
+      "name": "sourcesElements",
+      "attributes": {
+        "org.gradle.category": "documentation",
+        "org.gradle.dependency.bundling": "external",
+        "org.gradle.docstype": "sources",
+        "org.gradle.usage": "java-runtime"
+      },
+      "files": [
+        {
+          "name": "example-1.0-sources.jar",
+          "url": "example-1.0-sources.jar",
+          "size": 2002,
+          "sha512": "656e2b9daa2cce7bc88c8c647c83c1cabec73349bb25a1ed1f0d2724527fbbadd895ed5da53cee8e3516fc26b627be9e291668511a6196d2bf6e9804c441d2ba",
+          "sha256": "dd5da2763bba00328a36337dd08f73849b7d16d0214701bfe17b73e41d89faf7",
+          "sha1": "bbab2f32e16cc485c84e30a0918089b7f34a6fd8",
+          "md5": "c8e911f188115dd20539dcd4b1bccbc0"
+        }
+      ]
+    }
+  ]
+}

--- a/src/test/resources/with-sources/pom.xml
+++ b/src/test/resources/with-sources/pom.xml
@@ -1,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <!-- do_not_remove: published-with-gradle-metadata -->
+  <modelVersion>4.0.0</modelVersion>
+  <groupId>org.testing</groupId>
+  <artifactId>example</artifactId>
+  <version>1.0</version>
+  <packaging>jar</packaging>
+  <name>Test GMM Mojo</name>
+  <properties>
+    <maven.compiler.release>8</maven.compiler.release>
+    <project.build.outputTimestamp>2023-01-01T00:00:00Z</project.build.outputTimestamp>
+  </properties>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-source-plugin</artifactId>
+        <executions>
+          <execution>
+            <id>attach-sources</id>
+            <goals>
+              <goal>jar</goal>
+            </goals>
+          </execution>
+        </executions>
+      </plugin>
+      <plugin>
+        <groupId>org.gradlex</groupId>
+        <artifactId>gradle-module-metadata-maven-plugin</artifactId>
+        <executions>
+          <execution>
+            <goals>
+              <goal>gmm</goal>
+            </goals>
+          </execution>
+        </executions>
+        <configuration>
+          <platformDependencies>
+            <dependency>
+              <groupId>org.junit</groupId>
+              <artifactId>junit-bom</artifactId>
+              <version>5.6.0</version>
+            </dependency>
+          </platformDependencies>
+        </configuration>
+      </plugin>
+    </plugins>
+  </build>
+</project>

